druid-shell/src/platform/x11/window.rs

```diff
@@ -0,0 +1,407 @@
+        log::warn!("WindowBuilder::show_titlebar not implemented for x11")
+    }
+
+    pub fn set_title<S: Into<String>>(&mut self, title: S) {
+        self.title = title.into();
+    }
+
+    pub fn set_menu(&mut self, _menu: Menu) {
+        // TODO(x11/menus): implement WindowBuilder::set_menu (currently a no-op)
+    }
+
+    // TODO(x11/menus): make menus if requested
+    pub fn build(self) -> Result<WindowHandle, Error> {
+        let conn = Application::get_connection();
+        let screen_num = Application::get_screen_num();
+        let window_id = conn.generate_id();
+        let setup = conn.get_setup();
+        // TODO(x11/errors): Don't unwrap for screen or visual_type?
+        let screen = setup.roots().nth(screen_num as usize).unwrap();
+        let visual_type = get_visual_from_screen(&screen).unwrap();
+        let visual_id = visual_type.visual_id();
+
+        let cw_values = [
+            (xcb::CW_BACK_PIXEL, screen.white_pixel()),
+            (
+                xcb::CW_EVENT_MASK,
+                xcb::EVENT_MASK_EXPOSURE
+                    | xcb::EVENT_MASK_KEY_PRESS
+                    | xcb::EVENT_MASK_KEY_RELEASE
+                    | xcb::EVENT_MASK_BUTTON_PRESS
+                    | xcb::EVENT_MASK_BUTTON_RELEASE,
+            ),
+        ];
+
+        // Create the actual window
+        xcb::create_window(
+            // Connection to the X server
+            &conn,
+            // Window depth
+            XCB_COPY_FROM_PARENT.try_into().unwrap(),
+            // The new window's ID
+            window_id,
+            // Parent window of this new window
+            // TODO(#468): either `screen.root()` (no parent window) or pass parent here to attach
+            screen.root(),
+            // X-coordinate of the new window
+            0,
+            // Y-coordinate of the new window
+            0,
+            // Width of the new window
+            // TODO(x11/dpi_scaling): figure out DPI scaling
+            self.size.width as u16,
+            // Height of the new window
+            // TODO(x11/dpi_scaling): figure out DPI scaling
+            self.size.height as u16,
+            // Border width
+            0,
+            // Window class type
+            xcb::WINDOW_CLASS_INPUT_OUTPUT as u16,
+            // Visual ID
+            visual_id,
+            // Window properties mask
+            &cw_values,
+        );
+
+        // Set window title
+        xcb::change_property(
+            &conn,
+            xcb::PROP_MODE_REPLACE as u8,
+            window_id,
+            xcb::ATOM_WM_NAME,
+            xcb::ATOM_STRING,
+            8,
+            self.title.as_bytes(),
+        );
+
+        conn.flush();
+
+        let xwindow = XWindow::new(window_id, self.handler.unwrap(), self.size);
+        Application::add_xwindow(window_id, xwindow);
+
+        Ok(WindowHandle::new(window_id))
+    }
+}
+
+// X11-specific event handling and window drawing (etc.)
+pub struct XWindow {
```

Comment:
this shouldn't need to be public; at most it should be `pub(crate)`.

Reply:
Done.